### PR TITLE
Integrating external api

### PR DESCRIPTION
**Pull request by Sspadea**
## Summary

- Openlibrary external api integrated.
- Gem HTTParty is not working, finally we decided to use gem faraday.
- The endpoint is already returning the information of any book using its isbn.

## Trello Card

https://trello.com/b/Y5WdkBZf/training-ror-sebastian-spadea

## Screenshots or Screencasts

Captura de Pantalla 2021-06-10 a la(s) 11 41 56

### CODE REVIEW

File: app/controllers/api/v1/books_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Api
+  module V1
+    class BooksController < ApplicationController
+      def info
+        open_library = OpenLibrary.new(params[:isbn])
+        book = open_library.fetch_data
```
Comment (blacksam07) [inline app/controllers/api/v1/books_controller.rb]:
```suggestion
        book = OpenLibrary.new(params[:isbn]).fetch_data
```

File: app/controllers/api/v1/books_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        respond_to do |format|
+          if book != 'not found'
+            format.json { render json: { status: 'Success', message: 'Loaded book', data: book }, status: :ok }
```
Comment (blacksam07) [inline app/controllers/api/v1/books_controller.rb]: Use only a `render` method is more clear 
Success request
```ruby
render json: book, status: :ok
```
Wrong request like 
```ruby
render json: { message: 'Lorem itsu' }, status: :unprocessable_entity
```

File: app/services/application_service.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class ApplicationService
```
Comment (blacksam07) [inline app/services/application_service.rb]: Useless file

File: app/services/open_library.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      response = response[@isbn.to_s].slice('title', 'subtitle', 'authors', 'number_of_pages')
+      response['isbn'] = @isbn
+      response['authors'] = take_authors(response['authors'])
+      response
+    else
+      response
+    end
+  end
+
+  def books
```
Comment (blacksam07) [inline app/services/open_library.rb]: rename this method like `find_book`

File: app/services/open_library.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    uri = "https://openlibrary.org/api/books?bibkeys=#{@isbn}&format=json&jscmd=data"
```
Comment (blacksam07) [inline app/services/open_library.rb]: Use credentials to save the base `https://openlibrary.org/api` and merge with `/books` inside method, and remember to use the query params as JSON, you should define inside the initializer

```ruby
@options = { query: { bibkeys: "ISBN:#{isbn}", format: 'json', jscmd: 'data' } }

Faraday.get("#{base_url}/books", @options,
```

File: app/services/open_library.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    res = Faraday.get(uri)
+    res.body != '{}' ? res.body : 'not found'
```
Comment (blacksam07) [inline app/services/open_library.rb]: Use Json parse in body response and return, empty json or json with data

File: app/services/open_library.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    @isbn = isbn
+  end
+
+  def fetch_data
+    response = books
+    if response != 'not found'
+      response = JSON.parse(response)
+      response = response[@isbn.to_s].slice('title', 'subtitle', 'authors', 'number_of_pages')
+      response['isbn'] = @isbn
+      response['authors'] = take_authors(response['authors'])
```
Comment (blacksam07) [inline app/services/open_library.rb]: move this logic to external method

File: config/routes.rb [hunk 8]
```diff
@@ -1,4 +1,9 @@
 Rails.application.routes.draw do
   mount_devise_token_auth_for 'User', at: 'auth'
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  namespace :api do
+    namespace :v1 do
+      get "/books/:isbn" => "books#info"
```
Comment (blacksam07) [inline config/routes.rb]: use resources 

```ruby
resources :open_library, only: :show, param: :isbn
```

**Review by blacksam07 — CHANGES_REQUESTED** [review]

File: app/assets/stylesheets/open_library.scss [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the open_library controller here.
```
Comment (blacksam07) [inline app/assets/stylesheets/open_library.scss]: Useless file

File: app/helpers/open_library_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module OpenLibraryHelper
```
Comment (blacksam07) [inline app/helpers/open_library_helper.rb]: Useless file

File: app/services/open_library.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
+class OpenLibrary
+  require 'faraday'
+  require 'json'
+
+  def initialize(isbn)
+    @isbn = isbn
+    @options = { bibkeys: "ISBN:#{isbn}", format: 'json', jscmd: 'data' }
+    @base_url = 'https://openlibrary.org/api'
```
Comment (blacksam07) [inline app/services/open_library.rb]: Use credentials to save the url https://blog.saeloun.com/2019/10/10/rails-6-adds-support-for-multi-environment-credentials.html#in-rails-6

**Review by blacksam07 — APPROVED** [review]

File: app/controllers/open_library_controller.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class OpenLibraryController < ApplicationController
+  def show
+    book = OpenLibrary.new(params[:isbn]).fetch_data
+    if book.empty?
+      render json: book, status: :unprocessable_entity
+    else
+      render json: book, status: :ok
```
Comment (redwarewolf) [inline app/controllers/open_library_controller.rb]: The status and response isn't right. When you can't find a book it's 404 resource not found.
You can read more about it here: https://github.com/Wolox/tech-guides/blob/master/backend/backend-standard-and-best-practices.md#32--error-forwarding

Basically the response should be: { error: { some error and message}, status: status_code }

File: app/services/open_library.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    response = find_book
+    unless response.empty?
+      response = make_book_hash(response)
+    end
+    response
+  end
+
+  def find_book
+    url = "#{Rails.application.credentials.config[:url]}/books"
+    byebug
```
Comment (redwarewolf) [inline app/services/open_library.rb]: Oops 👀

File: app/services/open_library.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  def fetch_data
+    response = find_book
+    unless response.empty?
+      response = make_book_hash(response)
+    end
+    response
+  end
+
+  def find_book
+    url = "#{Rails.application.credentials.config[:url]}/books"
```
Comment (redwarewolf) [inline app/services/open_library.rb]: This should be a constant outside the method

**Review by redwarewolf — CHANGES_REQUESTED** [review]

File: app/services/open_library.rb (hunk 13 above)
Comment (redwarewolf) [reply app/services/open_library.rb]: @blacksam07 :shame: jajajajaja

**Review by redwarewolf — APPROVED** [review]

**Pull request merged**